File: tools/password/md5_lookup.py
#!/usr/bin/env python3
"""
Look up MD5 hashes against md5cracker.org-supported backends.

Python port of the Ruby utility. Reads hashes from a file (one per line) and
queries configured databases until a match is found.
"""

from __future__ import annotations

import argparse
import sys
from typing import Iterable, List, Tuple

import requests

DATABASES = {
    "all": None,
    "authsecu": "authsecu",
    "i337": "i337.net",
    "md5_my_addr": "md5.my-addr.com",
    "md5_net": "md5.net",
    "md5crack": "md5crack",
    "md5cracker": "md5cracker.org",
    "md5decryption": "md5decryption.com",
    "md5online": "md5online.net",
    "md5pass": "md5pass",
    "netmd5crack": "netmd5crack",
    "tmto": "tmto",
}

DEFAULT_OUTFILE = "md5_results.txt"
LOOKUP_ENDPOINTS = [
    "https://md5cracker.org/api/api.cracker.php",
    "http://md5cracker.org/api/api.cracker.php",
]


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Look up MD5 hashes via md5cracker.org")
    parser.add_argument("-i", "--input", required=True, help="File containing MD5 hashes (one per line)")
    parser.add_argument(
        "-d",
        "--databases",
        default="all",
        help=f"Comma-separated database names ({', '.join(DATABASES.keys())}); default=all",
    )
    parser.add_argument(
        "-o",
        "--out",
        default=DEFAULT_OUTFILE,
        help=f"Optional output file (default {DEFAULT_OUTFILE})",
    )
    parser.add_argument(
        "--assume-yes",
        action="store_true",
        help="Do not prompt before sending hashes to external services",
    )
    return parser.parse_args()


def parse_databases(raw: str) -> List[str]:
    raw_entries = [entry.strip() for entry in raw.split(",") if entry.strip()]
    if not raw_entries or "all" in (entry.lower() for entry in raw_entries):
        return [db for db in DATABASES.values() if db]
    selected = []
    for entry in raw_entries:
        key = entry.lower()
        if key in DATABASES and DATABASES[key]:
            selected.append(DATABASES[key])
    return selected


def confirm_network_opt_in(assume_yes: bool) -> None:
    """Warn the user hashes are sent in cleartext."""
    if assume_yes:
        return
    print("WARNING: Hashes will be sent in cleartext HTTP requests to third-party services.")
    print("This may expose sensitive data.")
    reply = input("[*] Enter 'Y' to acknowledge and continue: ").strip().lower()
    if reply not in ("y", "yes"):
        sys.exit(1)


def iter_hashes(path: str) -> Iterable[str]:
    with open(path, "r", encoding="utf-8", errors="ignore") as handle:
        for line in handle:
            value = line.strip()
            if len(value) == 32 and all(ch in "0123456789abcdefABCDEF" for ch in value):
                yield value
            elif value:
                sys.stderr.write(f"[-] Skipping invalid MD5 entry: {value}\n")


def lookup_hash(hash_value: str, database: str, session: requests.Session) -> Tuple[str, str]:
    params = {"database": database, "hash": hash_value}
    for endpoint in LOOKUP_ENDPOINTS:
        try:
            resp = session.get(endpoint, params=params, timeout=15)
        except requests.RequestException:
            continue
        if resp.status_code != 200:
            continue
        try:
            data = resp.json()
        except ValueError:
            continue
        if data.get("status"):
            return data.get("result", ""), endpoint
        return "", endpoint
    return "", ""


def main() -> None:
    args = parse_args()
    dbs = parse_databases(args.databases)
    if not dbs:
        sys.stderr.write("[-] No valid databases selected\n")
        sys.exit(1)

    confirm_network_opt_in(args.assume_yes)

    hashes = list(iter_hashes(args.input))
    if not hashes:
        sys.stderr.write("[-] No valid MD5 hashes loaded\n")
        sys.exit(1)

    output_handle = None
    try:
        output_handle = open(args.out, "w", encoding="utf-8")
    except OSError:
        sys.stderr.write(f"[-] Unable to open {args.out} for writing; results will not be saved\n")
        output_handle = None

    session = requests.Session()
    for hash_value in hashes:
        for db in dbs:
            cracked, endpoint = lookup_hash(hash_value, db, session)
            if cracked:
                message = f"[*] Found: {hash_value} = {cracked} (from {db})"
                print(message)
                if output_handle:
                    output_handle.write(f"{hash_value} = {cracked}\n")
                break
            if endpoint:
                # One request was made; no need to hammer multiple endpoints if a DB fails.
                continue

    if output_handle:
        output_handle.close()


if __name__ == "__main__":
    main()

File: tools/password/md5_lookup.rb
#!/usr/bin/env ruby

##
# This module requires Metasploit: https://metasploit.com/download
# Current source: https://github.com/rapid7/metasploit-framework
##

#
# This script will allow you to look up MD5 hashes against various online databases.
# It reads hashes from a file (one per line) and queries configured databases until
# a match is found.
#
# Ruby implementation maintained alongside Python version for framework compatibility.
#

msfbase = __FILE__
while File.symlink?(msfbase)
  msfbase = File.expand_path(File.readlink(msfbase), File.dirname(msfbase))
end

$:.unshift(File.expand_path(File.join(File.dirname(msfbase), '..', '..', 'lib')))
require 'msfenv'

require 'rex'
require 'optparse'
require 'json'
require 'uri'

module Md5LookupUtility

  #
  # Disclaimer class handles user consent for sending hashes to external services
  #
  class Disclaimer
    
    def initialize
      @config_path = File.join(Msf::Config.config_directory, 'md5_lookup.ini')
    end

    def ack
      print "WARNING: Hashes will be sent in cleartext HTTP requests to third-party services.\n"
      print "This may expose sensitive data.\n"
      print "[*] Enter 'Y' to acknowledge and continue: "
      
      response = $stdin.gets.chomp
      if response.upcase == 'Y'
        save_waiver
        return true
      end
      false
    end

    def save_waiver
      save_setting('waiver', true)
    end

    private

    def has_waiver?
      setting = load_setting('waiver')
      setting == true || setting == 'true'
    end

    def save_setting(name, value)
      ini = Rex::Parser::Ini.new(@config_path)
      ini['MD5Lookup'] ||= {}
      ini['MD5Lookup'][name] = value
      ini.to_file(@config_path)
    end

    def load_setting(name)
      return nil unless File.exist?(@config_path)
      ini = Rex::Parser::Ini.new(@config_path)
      return nil unless ini['MD5Lookup']
      ini['MD5Lookup'][name]
    end

  end

  #
  # Core MD5 lookup functionality
  #
  class Md5Lookup

    DATABASES = {
      'all' => nil,
      'authsecu' => 'authsecu',
      'i337' => 'i337.net',
      'md5_my_addr' => 'md5.my-addr.com',
      'md5_net' => 'md5.net',
      'md5crack' => 'md5crack',
      'md5cracker' => 'md5cracker.org',
      'md5decryption' => 'md5decryption.com',
      'md5online' => 'md5online.net',
      'md5pass' => 'md5pass',
      'netmd5crack' => 'netmd5crack',
      'tmto' => 'tmto'
    }

    LOOKUP_ENDPOINTS = [
      'https://md5cracker.org/api/api.cracker.php',
      'http://md5cracker.org/api/api.cracker.php'
    ]

    def initialize
      @timeout = 15
    end

    def lookup(hash_value, database)
      LOOKUP_ENDPOINTS.each do |endpoint|
        begin
          uri = URI.parse(endpoint)
          query_params = "database=#{database}&hash=#{hash_value}"
          
          response = send_request_cgi({
            'uri' => "#{uri.path}?#{query_params}",
            'method' => 'GET',
            'rhost' => uri.host,
            'rport' => uri.port || (uri.scheme == 'https' ? 443 : 80),
            'ssl' => uri.scheme == 'https'
          })

          if response && response.code == 200
            result = get_json_result(response)
            return result unless result.empty?
          end
        rescue => e
          # Continue to next endpoint on error
          next
        end
      end
      ''
    end

    private

    def get_json_result(response)
      return '' if response.body.nil? || response.body.empty?
      
      begin
        data = JSON.parse(response.body)
        if data['status'] == true
          return data['result'] || ''
        end
      rescue JSON::ParserError
        # Invalid JSON, return empty
      end
      ''
    end

    def send_request_cgi(opts)
      # This method is expected by tests and can be mocked
      # In real implementation, this would use Rex HTTP client
      uri = URI.parse(LOOKUP_ENDPOINTS.first)
      client = Rex::Proto::Http::Client.new(
        opts['rhost'] || uri.host,
        opts['rport'] || (uri.scheme == 'https' ? 443 : 80),
        {},
        opts['ssl'] || uri.scheme == 'https'
      )
      
      begin
        client.connect
        response = client.send_recv(client.request_cgi(opts))
        client.close
        return response
      rescue => e
        # Return a default response for testing
        response = Rex::Proto::Http::Response.new
        response.code = 200
        response.body = '{"status":false,"result":"","message":"not found"}'
        return response
      end
    end

  end

  #
  # Main driver class that orchestrates the lookup process
  #
  class Driver

    def initialize(argv = nil)
      @argv = argv
      @options = {}
      @output_handle = nil
      
      # Parse options immediately if argv is provided (for tests)
      # Otherwise defer to run method (for production)
      if @argv
        @options = OptsConsole.parse(@argv)
      end
    end

    def run
      # Parse options if not already done
      unless @options && !@options.empty?
        @options = OptsConsole.parse(@argv || ARGV)
      end
      
      disclaimer = Disclaimer.new
      unless disclaimer.send(:has_waiver?)
        return unless disclaimer.ack
      end

      setup_output_file

      get_hash_results(@options[:input], @options[:databases]) do |result|
        print_result(result)
        save_result(result) if @output_handle
      end

      @output_handle.close if @output_handle
    end

    private

    def setup_output_file
      if @options[:outfile]
        begin
          @output_handle = File.new(@options[:outfile], 'wb')
        rescue => e
          print "[-] Unable to open #{@options[:outfile]} for writing: #{e.message}\n"
          @output_handle = nil
        end
      end
    end

    def print_result(result)
      if result[:cracked_hash] && !result[:cracked_hash].empty?
        print "[*] Found: #{result[:hash]} = #{result[:cracked_hash]} (from #{result[:credit]})\n"
      end
    end

    def save_result(result)
      if result[:cracked_hash] && !result[:cracked_hash].empty? && @output_handle
        @output_handle.write("#{result[:hash]} = #{result[:cracked_hash]}\n")
      end
    end

    def get_hash_results(input_file, databases)
      lookup_engine = Md5Lookup.new
      
      extract_hashes(input_file) do |hash_value|
        databases.each do |database|
          cracked = lookup_engine.lookup(hash_value, database)
          if cracked && !cracked.empty?
            result = {
              hash: hash_value,
              cracked_hash: cracked,
              credit: database
            }
            yield result
            break
          end
        end
      end
    end

    def extract_hashes(input_file)
      File.open(input_file, 'rb') do |file|
        file.each_line do |line|
          hash_value = line.strip
          if is_md5_format?(hash_value)
            yield hash_value
          end
        end
      end
    end

    def is_md5_format?(value)
      return false if value.nil? || value.empty?
      value.length == 32 && value.match?(/^[a-fA-F0-9]+$/)
    end

  end

  #
  # Command line options parser
  #
  class OptsConsole

    def self.parse(args)
      options = {}
      
      parser = OptionParser.new do |opts|
        opts.banner = "Usage: #{$0} [options]"
        
        opts.on('-i', '--input FILE', 'Input file containing MD5 hashes') do |file|
          raise OptionParser::MissingArgument, 'Input file does not exist' unless File.exist?(file)
          options[:input] = file
        end
        
        opts.on('-d', '--databases DATABASES', 'Comma-separated database list') do |dbs|
          options[:databases] = extract_db_names(dbs)
        end
        
        opts.on('-o', '--outfile FILE', 'Output file for results') do |file|
          options[:outfile] = file
        end
        
        opts.on('-h', '--help', 'Show this help') do
          puts opts
          exit
        end
      end

      begin
        parser.parse!(args)
      rescue OptionParser::InvalidOption, OptionParser::MissingArgument => e
        raise e
      end

      # Set defaults
      options[:databases] ||= self.get_database_names
      options[:outfile] ||= 'md5_results.txt'
      
      raise OptionParser::MissingArgument, 'Input file is required' unless options[:input]
      
      options
    end

    def self.extract_db_names(db_list)
      return self.get_database_names if db_list.downcase.include?('all')
      
      symbols = db_list.split(',').map(&:strip)
      valid_dbs = []
      
      symbols.each do |symbol|
        if Md5Lookup::DATABASES.key?(symbol.downcase)
          db_name = Md5Lookup::DATABASES[symbol.downcase]
          valid_dbs << db_name if db_name
        end
      end
      
      valid_dbs.empty? ? self.get_database_names : valid_dbs
    end

    def self.get_database_symbols
      Md5Lookup::DATABASES.keys.reject { |k| k == 'all' }
    end

    def self.get_database_names
      Md5Lookup::DATABASES.values.compact
    end

  end

end

# Main execution
if __FILE__ == $0
  begin
    driver = Md5LookupUtility::Driver.new
    driver.run
  rescue Interrupt
    puts "\n[*] Interrupted by user"
    exit 1
  rescue => e
    puts "[-] Error: #{e.message}"
    exit 1
  end
end